File: ListCycle.java
// Time Complexity : O(n) n is the number of elements in the linked list
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach



public class ListCycle {
    ListCycle(String s){ n=s;}
    String n;
    public ListNode detectCycle(ListNode head) {
        HashSet<ListNode> hm = new HashSet<>();
        while(true){
            if(head==null){
                return null;
            }
            if(hm.contains(head)){
                return head;
            }
            else{
                hm.add(head);
            }
            head = head.next;
        }
    }

    public static void main(String[] args){
        ListCycle obj = new ListCycle();

    }
}

File: ListCycle.cs
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode DetectCycle(ListNode head) {
        if(head == null) return null;
        ListNode fast = head;
        ListNode slow = head;
        bool flag = false;
        while(fast != null && fast.next != null)
        {
            slow = slow.next;
            fast = fast.next.next;
            if(slow == fast)
            {
                flag = true;
                break;
            }
        }
        if(!flag) return null;  
        slow = head;      
        while(slow != fast)
        {
            slow = slow.next;
            fast = fast.next;
        }
        return slow;
    }
}
